File: Backend/Services/Branch/PendingOrders/PendingOrdersService.cs
using Backend.Data.Branch;
using Backend.Models.DTOs.Branch.PendingOrders;
using Backend.Models.DTOs.Shared;
using Backend.Models.Entities.Branch;
using Backend.Models.Enums;
using Backend.Utilities;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;

namespace Backend.Services.Branch.PendingOrders;

/// <summary>
/// Service for managing pending orders
/// </summary>
public class PendingOrdersService : IPendingOrdersService
{
    private readonly BranchDbContext _context;
    private readonly ILogger<PendingOrdersService> _logger;

    public PendingOrdersService(
        BranchDbContext context,
        ILogger<PendingOrdersService> logger)
    {
        _context = context;
        _logger = logger;
    }

    public async Task<ApiResponse<PendingOrderDto>> CreatePendingOrderAsync(
        CreatePendingOrderDto dto,
        string userId,
        string username)
    {
        try
        {
            // Validate items
            if (dto.Items == null || !dto.Items.Any())
            {
                return ApiResponse<PendingOrderDto>.ErrorResponse("At least one item is required");
            }

            // Generate order number
            var orderNumber = OrderNumberGenerator.GenerateOrderNumber();

            // Create pending order entity
            var pendingOrder = new PendingOrder
            {
                Id = Guid.NewGuid(),
                OrderNumber = orderNumber,
                CustomerName = dto.CustomerName,
                CustomerPhone = dto.CustomerPhone,
                CustomerId = dto.CustomerId,
                TableId = dto.TableId,
                TableNumber = dto.TableNumber,
                GuestCount = dto.GuestCount,
                Subtotal = dto.Subtotal,
                TaxAmount = dto.TaxAmount,
                DiscountAmount = dto.DiscountAmount,
                TotalAmount = dto.TotalAmount,
                Notes = dto.Notes,
                OrderType = dto.OrderType,
                Status = dto.Status,
                CreatedAt = DateTime.UtcNow,
                ExpiresAt = DateTime.UtcNow.AddHours(24), // Expire after 24 hours
                CreatedByUserId = userId,
                CreatedByUsername = username,
                Items = dto.Items.Select(item => new PendingOrderItem
                {
                    Id = Guid.NewGuid(),
                    ProductId = item.ProductId,
                    ProductName = item.ProductName,
                    ProductSku = item.ProductSku,
                    UnitPrice = item.UnitPrice,
                    Quantity = item.Quantity,
                    Discount = item.Discount,
                    TotalPrice = item.TotalPrice,
                    Notes = item.Notes
                }).ToList()
            };

            _context.PendingOrders.Add(pendingOrder);
            await _context.SaveChangesAsync();

            _logger.LogInformation(
                "Pending order {OrderNumber} created by user {UserId}",
                orderNumber, userId);

            return ApiResponse<PendingOrderDto>.SuccessResponse(
                MapToDto(pendingOrder),
                "Pending order created successfully");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error creating pending order for user {UserId}", userId);
            return ApiResponse<PendingOrderDto>.ErrorResponse(
                "Failed to create pending order");
        }
    }

    public async Task<ApiResponse<PaginationResponse<PendingOrderDto>>> GetPendingOrdersAsync(
        PendingOrderStatus? status = null,
        string? createdBy = null,
        OrderType? orderType = null,
        string? search = null,
        int page = 1,
        int pageSize = 10,
        bool isManager = false,
        string? currentUserId = null)
    {
        try
        {
            var query = _context.PendingOrders
                .Include(po => po.Items)
                .AsQueryable();

            // Role-based filtering: Cashiers can only see their own orders
            if (!isManager && !string.IsNullOrEmpty(currentUserId))
            {
                query = query.Where(po => po.CreatedByUserId == currentUserId);
            }
            // Managers can filter by specific cashier if provided
            else if (isManager && !string.IsNullOrEmpty(createdBy))
            {
                query = query.Where(po => po.CreatedByUserId == createdBy);
            }

            // Filter by status
            if (status.HasValue)
            {
                query = query.Where(po => po.Status == status.Value);
            }

            // Filter by order type
            if (orderType.HasValue)
            {
                query = query.Where(po => po.OrderType == orderType.Value);
            }

            // Search by customer name, phone, or order number
            if (!string.IsNullOrWhiteSpace(search))
            {
                var searchLower = search.ToLower();
                query = query.Where(po =>
                    (po.CustomerName != null && po.CustomerName.ToLower().Contains(searchLower)) ||
                    (po.CustomerPhone != null && po.CustomerPhone.Contains(search)) ||
                    po.OrderNumber.ToLower().Contains(searchLower));
            }

            // Get total count
            var totalCount = await query.CountAsync();

            // Apply pagination
            var items = await query
                .OrderByDescending(po => po.CreatedAt)
                .Skip((page - 1) * pageSize)
                .Take(pageSize)
                .ToListAsync();

            var paginationResponse = new PaginationResponse<PendingOrderDto>
            {
                Items = items.Select(MapToDto).ToList(),
                TotalCount = totalCount,
                Page = page,
                PageSize = pageSize,
                TotalPages = (int)Math.Ceiling(totalCount / (double)pageSize)
            };

            return ApiResponse<PaginationResponse<PendingOrderDto>>.SuccessResponse(
                paginationResponse);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error retrieving pending orders");
            return ApiResponse<PaginationResponse<PendingOrderDto>>.ErrorResponse(
                "Failed to retrieve pending orders");
        }
    }

    public async Task<ApiResponse<PendingOrderDto>> GetPendingOrderByIdAsync(
        Guid id,
        string userId,
        bool isManager)
    {
        try
        {
            var pendingOrder = await _context.PendingOrders
                .Include(po => po.Items)
                .FirstOrDefaultAsync(po => po.Id == id);

            if (pendingOrder == null)
            {
                return ApiResponse<PendingOrderDto>.ErrorResponse("Pending order not found");
            }

            // Check permission: Cashiers can only view their own orders
            if (!isManager && pendingOrder.CreatedByUserId != userId)
            {
                return ApiResponse<PendingOrderDto>.ErrorResponse(
                    "You don't have permission to view this order");
            }

            return ApiResponse<PendingOrderDto>.SuccessResponse(MapToDto(pendingOrder));
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error retrieving pending order {OrderId}", id);
            return ApiResponse<PendingOrderDto>.ErrorResponse(
                "Failed to retrieve pending order");
        }
    }

    public async Task<ApiResponse<PendingOrderDto>> UpdatePendingOrderAsync(
        Guid id,
        UpdatePendingOrderDto dto,
        string userId,
        bool isManager)
    {
        try
        {
            var pendingOrder = await _context.PendingOrders
                .Include(po => po.Items)
                .FirstOrDefaultAsync(po => po.Id == id);

            if (pendingOrder == null)
            {
                return ApiResponse<PendingOrderDto>.ErrorResponse("Pending order not found");
            }

            // Check permission
            if (!isManager && pendingOrder.CreatedByUserId != userId)
            {
                return ApiResponse<PendingOrderDto>.ErrorResponse(
                    "You don't have permission to update this order");
            }

            // Update fields (only if provided)
            if (dto.CustomerName != null)
                pendingOrder.CustomerName = dto.CustomerName;

            if (dto.CustomerPhone != null)
                pendingOrder.CustomerPhone = dto.CustomerPhone;

            if (dto.CustomerId.HasValue)
                pendingOrder.CustomerId = dto.CustomerId;

            if (dto.TableId.HasValue)
                pendingOrder.TableId = dto.TableId;

            if (dto.TableNumber != null)
                pendingOrder.TableNumber = dto.TableNumber;

            if (dto.GuestCount.HasValue)
                pendingOrder.GuestCount = dto.GuestCount;

            if (dto.Notes != null)
                pendingOrder.Notes = dto.Notes;

            if (dto.OrderType.HasValue)
                pendingOrder.OrderType = dto.OrderType.Value;

            if (dto.Status.HasValue)
                pendingOrder.Status = dto.Status.Value;

            // Update items if provided
            if (dto.Items != null && dto.Items.Any())
            {
                // Remove old items
                _context.PendingOrderItems.RemoveRange(pendingOrder.Items);

                // Add new items
                pendingOrder.Items = dto.Items.Select(item => new PendingOrderItem
                {
                    Id = Guid.NewGuid(),
                    PendingOrderId = pendingOrder.Id,
                    ProductId = item.ProductId,
                    ProductName = item.ProductName,
                    ProductSku = item.ProductSku,
                    UnitPrice = item.UnitPrice,
                    Quantity = item.Quantity,
                    Discount = item.Discount,
                    TotalPrice = item.TotalPrice,
                    Notes = item.Notes
                }).ToList();
            }

            // Update totals if provided
            if (dto.Subtotal.HasValue)
                pendingOrder.Subtotal = dto.Subtotal.Value;

            if (dto.TaxAmount.HasValue)
                pendingOrder.TaxAmount = dto.TaxAmount.Value;

            if (dto.DiscountAmount.HasValue)
                pendingOrder.DiscountAmount = dto.DiscountAmount.Value;

            if (dto.TotalAmount.HasValue)
                pendingOrder.TotalAmount = dto.TotalAmount.Value;

            pendingOrder.UpdatedAt = DateTime.UtcNow;

            await _context.SaveChangesAsync();

            _logger.LogInformation(
                "Pending order {OrderNumber} updated by user {UserId}",
                pendingOrder.OrderNumber, userId);

            return ApiResponse<PendingOrderDto>.SuccessResponse(
                MapToDto(pendingOrder),
                "Pending order updated successfully");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error updating pending order {OrderId}", id);
            return ApiResponse<PendingOrderDto>.ErrorResponse(
                "Failed to update pending order");
        }
    }

    public async Task<ApiResponse<bool>> DeletePendingOrderAsync(
        Guid id,
        string userId,
        bool isManager)
    {
        try
        {
            var pendingOrder = await _context.PendingOrders
                .FirstOrDefaultAsync(po => po.Id == id);

            if (pendingOrder == null)
            {
                return ApiResponse<bool>.ErrorResponse("Pending order not found");
            }

            // Check permission
            if (!isManager && pendingOrder.CreatedByUserId != userId)
            {
                return ApiResponse<bool>.ErrorResponse(
                    "You don't have permission to delete this order");
            }

            _context.PendingOrders.Remove(pendingOrder);
            await _context.SaveChangesAsync();

            _logger.LogInformation(
                "Pending order {OrderNumber} deleted by user {UserId}",
                pendingOrder.OrderNumber, userId);

            return ApiResponse<bool>.SuccessResponse(
                true,
                "Pending order deleted successfully");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error deleting pending order {OrderId}", id);
            return ApiResponse<bool>.ErrorResponse("Failed to delete pending order");
        }
    }

    public async Task<ApiResponse<RetrievePendingOrderDto>> RetrievePendingOrderAsync(
        Guid id,
        string userId,
        bool isManager)
    {
        try
        {
            var pendingOrder = await _context.PendingOrders
                .Include(po => po.Items)
                .FirstOrDefaultAsync(po => po.Id == id);

            if (pendingOrder == null)
            {
                return ApiResponse<RetrievePendingOrderDto>.ErrorResponse(
                    "Pending order not found");
            }

            // Check permission
            if (!isManager && pendingOrder.CreatedByUserId != userId)
            {
                return ApiResponse<RetrievePendingOrderDto>.ErrorResponse(
                    "You don't have permission to retrieve this order");
            }

            // Mark as retrieved
            pendingOrder.Status = PendingOrderStatus.Retrieved;
            pendingOrder.RetrievedAt = DateTime.UtcNow;
            pendingOrder.UpdatedAt = DateTime.UtcNow;

            await _context.SaveChangesAsync();

            _logger.LogInformation(
                "Pending order {OrderNumber} retrieved by user {UserId}",
                pendingOrder.OrderNumber, userId);

            var dto = MapToRetrieveDto(pendingOrder);

            return ApiResponse<RetrievePendingOrderDto>.SuccessResponse(
                dto,
                "Pending order retrieved successfully");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error retrieving pending order {OrderId}", id);
            return ApiResponse<RetrievePendingOrderDto>.ErrorResponse(
                "Failed to retrieve pending order");
        }
    }

    public async Task<ApiResponse<Guid>> ConvertToSaleAsync(Guid id, string userId)
    {
        try
        {
            var pendingOrder = await _context.PendingOrders
                .Include(po => po.Items)
                .FirstOrDefaultAsync(po => po.Id == id);

            if (pendingOrder == null)
            {
                return ApiResponse<Guid>.ErrorResponse("Pending order not found");
            }

            // TODO: Implement conversion to Sale
            // This would create a new Sale entity from the pending order
            // For now, just delete the pending order after marking as retrieved

            _context.PendingOrders.Remove(pendingOrder);
            await _context.SaveChangesAsync();

            _logger.LogInformation(
                "Pending order {OrderNumber} converted to sale by user {UserId}",
                pendingOrder.OrderNumber, userId);

            return ApiResponse<Guid>.SuccessResponse(
                Guid.NewGuid(), // TODO: Return actual Sale ID
                "Pending order converted to sale successfully");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error converting pending order {OrderId} to sale", id);
            return ApiResponse<Guid>.ErrorResponse("Failed to convert pending order to sale");
        }
    }

    public async Task<ApiResponse<PendingOrderStatsDto>> GetPendingOrderStatsAsync()
    {
        try
        {
            var pendingOrders = await _context.PendingOrders
                .Include(po => po.Items)
                .ToListAsync();

            var now = DateTime.UtcNow;

            var stats = new PendingOrderStatsDto
            {
                TotalPendingOrders = pendingOrders.Count,
                OrdersByStatus = pendingOrders
                    .GroupBy(po => po.Status)
                    .ToDictionary(g => g.Key, g => g.Count()),
                OrdersByUser = pendingOrders
                    .GroupBy(po => po.CreatedByUsername)
                    .ToDictionary(g => g.Key, g => g.Count()),
                OrdersByType = pendingOrders
                    .GroupBy(po => po.OrderType)
                    .ToDictionary(g => g.Key, g => g.Count()),
                TotalPendingValue = pendingOrders.Sum(po => po.TotalAmount),
                OrdersExpiringSoon = pendingOrders
                    .Count(po => po.ExpiresAt > now && po.ExpiresAt <= now.AddMinutes(30)),
                ExpiredOrders = pendingOrders.Count(po => po.ExpiresAt <= now),
                OldestPendingOrder = pendingOrders.Any()
                    ? pendingOrders.Min(po => po.CreatedAt)
                    : null,
                NewestPendingOrder = pendingOrders.Any()
                    ? pendingOrders.Max(po => po.CreatedAt)
                    : null,
                AveragePendingTimeMinutes = pendingOrders.Any()
                    ? pendingOrders.Average(po => (now - po.CreatedAt).TotalMinutes)
                    : 0
            };

            return ApiResponse<PendingOrderStatsDto>.SuccessResponse(stats);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error retrieving pending order stats");
            return ApiResponse<PendingOrderStatsDto>.ErrorResponse(
                "Failed to retrieve pending order statistics");
        }
    }

    public async Task<int> DeleteExpiredOrdersAsync()
    {
        try
        {
            var now = DateTime.UtcNow;
            var expiredOrders = await _context.PendingOrders
                .Where(po => po.ExpiresAt <= now)
                .ToListAsync();

            if (expiredOrders.Any())
            {
                _context.PendingOrders.RemoveRange(expiredOrders);
                await _context.SaveChangesAsync();

                _logger.LogInformation(
                    "Deleted {Count} expired pending orders",
                    expiredOrders.Count);
            }

            return expiredOrders.Count;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error deleting expired pending orders");
            return 0;
        }
    }

    // Helper methods for mapping entities to DTOs
    private static PendingOrderDto MapToDto(PendingOrder entity)
    {
        return new PendingOrderDto
        {
            Id = entity.Id,
            OrderNumber = entity.OrderNumber,
            CustomerName = entity.CustomerName,
            CustomerPhone = entity.CustomerPhone,
            CustomerId = entity.CustomerId,
            TableId = entity.TableId,
            TableNumber = entity.TableNumber,
            GuestCount = entity.GuestCount,
            Items = entity.Items.Select(item => new PendingOrderItemDto
            {
                Id = item.Id,
                ProductId = item.ProductId,
                ProductName = item.ProductName,
                ProductSku = item.ProductSku,
                UnitPrice = item.UnitPrice,
                Quantity = item.Quantity,
                Discount = item.Discount,
                TotalPrice = item.TotalPrice,
                Notes = item.Notes
            }).ToList(),
            Subtotal = entity.Subtotal,
            TaxAmount = entity.TaxAmount,
            DiscountAmount = entity.DiscountAmount,
            TotalAmount = entity.TotalAmount,
            Notes = entity.Notes,
            OrderType = entity.OrderType,
            Status = entity.Status,
            CreatedAt = entity.CreatedAt,
            UpdatedAt = entity.UpdatedAt,
            CreatedByUserId = entity.CreatedByUserId,
            CreatedByUsername = entity.CreatedByUsername,
            RetrievedAt = entity.RetrievedAt,
            ExpiresAt = entity.ExpiresAt
        };
    }

    private static RetrievePendingOrderDto MapToRetrieveDto(PendingOrder entity)
    {
        var dto = new RetrievePendingOrderDto
        {
            Id = entity.Id,
            OrderNumber = entity.OrderNumber,
            CustomerName = entity.CustomerName,
            CustomerPhone = entity.CustomerPhone,
            CustomerId = entity.CustomerId,
            TableId = entity.TableId,
            TableNumber = entity.TableNumber,
            GuestCount = entity.GuestCount,
            Items = entity.Items.Select(item => new PendingOrderItemDto
            {
                Id = item.Id,
                ProductId = item.ProductId,
                ProductName = item.ProductName,
                ProductSku = item.ProductSku,
                UnitPrice = item.UnitPrice,
                Quantity = item.Quantity,
                Discount = item.Discount,
                TotalPrice = item.TotalPrice,
                Notes = item.Notes
            }).ToList(),
            Subtotal = entity.Subtotal,
            TaxAmount = entity.TaxAmount,
            DiscountAmount = entity.DiscountAmount,
            TotalAmount = entity.TotalAmount,
            Notes = entity.Notes,
            OrderType = entity.OrderType,
            Status = entity.Status,
            CreatedAt = entity.CreatedAt,
            UpdatedAt = entity.UpdatedAt,
            CreatedByUserId = entity.CreatedByUserId,
            CreatedByUsername = entity.CreatedByUsername,
            RetrievedAt = entity.RetrievedAt,
            ExpiresAt = entity.ExpiresAt,
            WasRetrieved = entity.RetrievedAt.HasValue,
            RetrievalTimestamp = entity.RetrievedAt
        };

        return dto;
    }
}

File: frontend/services/pending-orders.service.ts
/**
 * Pending Orders Service
 * Handles all pending orders-related API operations
 */

import api, { apiHelpers } from './api';
import offlineSyncQueue from '@/lib/offline-sync';
import {
  ApiResponse,
  PaginationResponse,
  CreatePendingOrderDto,
  UpdatePendingOrderDto,
  PendingOrderDto,
  RetrievePendingOrderDto,
  PendingOrderStatsDto,
  PendingOrderStatus,
} from '@/types/api.types';

/**
 * Query parameters for listing pending orders
 */
export interface GetPendingOrdersParams {
  page?: number;
  pageSize?: number;
  status?: PendingOrderStatus;
  createdBy?: string;
  orderType?: number;
  tableNumber?: string;
  search?: string;
  sortBy?: string;
  sortOrder?: 'asc' | 'desc';
}

/**
 * Pending Orders Service - Handles all pending orders operations
 */
class PendingOrdersService {
  private basePath = '/api/v1/pending-orders';

  /**
   * Create a new pending order
   * Supports offline mode by queuing transaction for sync when back online
   * @param orderData - Pending order creation data
   * @returns Created pending order (or optimistic response if offline)
   */
  async createPendingOrder(
    orderData: CreatePendingOrderDto
  ): Promise<PendingOrderDto> {
    // Check if online
    const isOnline = typeof navigator !== 'undefined' ? navigator.onLine : true;

    // If offline, queue transaction for sync
    if (!isOnline) {
      try {
        // Get branch and user context from local storage or session
        const branchId = localStorage.getItem('branchId') || '';
        const userId = localStorage.getItem('userId') || '';

        // Queue transaction for offline sync
        const transactionId = await offlineSyncQueue.add({
          type: 'pending_order',
          timestamp: new Date(),
          branchId,
          userId,
          data: orderData,
        });

        // Return optimistic response
        const optimisticResponse: PendingOrderDto = {
          id: transactionId, // Use transaction ID as temporary ID
          orderNumber: `PO-OFFLINE-${Date.now()}`, // Temporary order number
          customerName: orderData.customerName,
          customerPhone: orderData.customerPhone,
          customerId: orderData.customerId,
          tableId: orderData.tableId,
          tableNumber: orderData.tableNumber?.toString(),
          guestCount: orderData.guestCount,
          items: orderData.items,
          subtotal: orderData.subtotal,
          taxAmount: orderData.taxAmount,
          discountAmount: orderData.discountAmount,
          totalAmount: orderData.totalAmount,
          notes: orderData.notes,
          orderType: orderData.orderType,
          status: orderData.status,
          createdAt: new Date().toISOString(),
          updatedAt: new Date().toISOString(),
          createdByUserId: userId,
          createdByUsername: localStorage.getItem('username') || 'Unknown',
          retrievedAt: undefined,
          expiresAt: new Date(Date.now() + 24 * 60 * 60 * 1000).toISOString(), // 24 hours
        };

        return optimisticResponse;
      } catch (error) {
        const errorMessage = apiHelpers.getErrorMessage(error);
        throw new Error(`Failed to queue pending order for offline sync: ${errorMessage}`);
      }
    }

    // Online - create pending order normally
    try {
      const response = await api.post<ApiResponse<PendingOrderDto>>(
        this.basePath,
        orderData
      );
      return response.data.data!;
    } catch (error) {
      const errorMessage = apiHelpers.getErrorMessage(error);
      throw new Error(`Failed to create pending order: ${errorMessage}`);
    }
  }

  /**
   * Get a list of pending orders with optional filtering and pagination
   * @param params - Query parameters for filtering and pagination
   * @returns Paginated list of pending orders
   */
  async getPendingOrders(
    params?: GetPendingOrdersParams
  ): Promise<ApiResponse<any>> {
    try {
      const queryString = params ? apiHelpers.buildQueryString(params) : '';
      const url = queryString ? `${this.basePath}?${queryString}` : this.basePath;

      // Backend returns ApiResponse with nested paginated data structure
      const response = await api.get<ApiResponse<any>>(url);
      return response.data;
    } catch (error) {
      const errorMessage = apiHelpers.getErrorMessage(error);
      throw new Error(`Failed to fetch pending orders: ${errorMessage}`);
    }
  }

  /**
   * Get a pending order by ID
   * @param id - Pending order ID
   * @returns Pending order details
   */
  async getPendingOrderById(id: string): Promise<PendingOrderDto> {
    try {
      const response = await api.get<ApiResponse<PendingOrderDto>>(
        `${this.basePath}/${id}`
      );
      return response.data.data!;
    } catch (error) {
      const errorMessage = apiHelpers.getErrorMessage(error);
      throw new Error(`Failed to fetch pending order: ${errorMessage}`);
    }
  }

  /**
   * Update a pending order
   * @param id - Pending order ID
   * @param orderData - Updated pending order data
   * @returns Updated pending order
   */
  async updatePendingOrder(
    id: string,
    orderData: UpdatePendingOrderDto
  ): Promise<PendingOrderDto> {
    try {
      const response = await api.put<ApiResponse<PendingOrderDto>>(
        `${this.basePath}/${id}`,
        orderData
      );
      return response.data.data!;
    } catch (error) {
      const errorMessage = apiHelpers.getErrorMessage(error);
      throw new Error(`Failed to update pending order: ${errorMessage}`);
    }
  }

  /**
   * Delete a pending order
   * @param id - Pending order ID
   * @returns Success response
   */
  async deletePendingOrder(id: string): Promise<void> {
    try {
      await api.delete(`${this.basePath}/${id}`);
    } catch (error) {
      const errorMessage = apiHelpers.getErrorMessage(error);
      throw new Error(`Failed to delete pending order: ${errorMessage}`);
    }
  }

  /**
   * Retrieve a pending order (marks it as Retrieved)
   * @param id - Pending order ID
   * @returns Retrieved pending order
   */
  async retrievePendingOrder(id: string): Promise<RetrievePendingOrderDto> {
    try {
      const response = await api.post<ApiResponse<RetrievePendingOrderDto>>(
        `${this.basePath}/${id}/retrieve`
      );
      return response.data.data!;
    } catch (error) {
      const errorMessage = apiHelpers.getErrorMessage(error);
      throw new Error(`Failed to retrieve pending order: ${errorMessage}`);
    }
  }

  /**
   * Convert a pending order to a completed sale
   * @param id - Pending order ID
   * @returns Created sale
   */
  async convertToSale(id: string): Promise<any> {
    try {
      const response = await api.post<ApiResponse<any>>(
        `${this.basePath}/${id}/convert-to-sale`
      );
      return response.data.data!;
    } catch (error) {
      const errorMessage = apiHelpers.getErrorMessage(error);
      throw new Error(`Failed to convert pending order to sale: ${errorMessage}`);
    }
  }

  /**
   * Get pending orders statistics (Manager only)
   * @returns Statistics about pending orders
   */
  async getStats(): Promise<PendingOrderStatsDto> {
    try {
      const response = await api.get<ApiResponse<PendingOrderStatsDto>>(
        `${this.basePath}/stats`
      );
      return response.data.data!;
    } catch (error) {
      const errorMessage = apiHelpers.getErrorMessage(error);
      throw new Error(`Failed to fetch pending orders stats: ${errorMessage}`);
    }
  }
}

// Export singleton instance
const pendingOrdersService = new PendingOrdersService();
export default pendingOrdersService;
